### ISSUE TITLE
[clang][analyzer] False positive core.NullDereference after updating list head through pointer-to-pointer backlink

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#215755]
**Description:**
The analyzer reports a null-pointer dereference on an impossible second loop iteration. The non-returning path in `reproducer()` expects a valid one-element list (`head -> [node] -> NULL`), and when updated inside `unlink_node()`, line `*node->previous = node->next;` is equivalent to `*(&head) = NULL`, so `head` becomes a `NULL` and the loop body execute only once.

**Reproducer:**
```c
#define NULL    ((void*)0)

void clang_analyzer_eval(int);
void clang_analyzer_numTimesReached(void);

typedef struct Node Node;
struct Node {
    Node *next;
    Node **previous;
};

static Node *head;

void unlink_node(Node *node)
{
    *node->previous = node->next;

    if (node->next != NULL)
        node->next->previous = node->previous;

    node->previous = NULL;
    node->next = NULL;
}

void reproducer(void)
{
    if (head == NULL || head->next != NULL || head->previous != &head)
        return;

    clang_analyzer_eval(head->previous == &head);

    while (head != 0) {
        clang_analyzer_numTimesReached();

        Node *node = head;
        unlink_node(node);

        clang_analyzer_eval(head == NULL);
    }
}
```

**Output:**
```sh
mre.c:18:21: warning: Dereference of null pointer (loaded from field 'previous') [core.NullDereference]
   18 |     *node->previous = node->next;
      |            ~~~~~~~~ ^
mre.c:32:5: warning: UNKNOWN [debug.ExprInspection]
   32 |     clang_analyzer_eval(head->previous == &head);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mre.c:35:2: warning: 2 [debug.ExprInspection]
   35 |         clang_analyzer_numTimesReached();
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mre.c:40:2: warning: FALSE [debug.ExprInspection]
   40 |         clang_analyzer_eval(head == NULL);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
However, the analyzer:
- reports UNKNOWN for `head->previous == &head` (should be TRUE);
- reports FALSE for `head == NULL` after unlinking (should be TRUE);
- reaches the loop body twice (should be only once);
- reports a null dereference during the second iteration.

**Clang version:**
```sh
clang version 24.0.0git (https://github.com/llvm/llvm-project.git d71427f70210606cc41a460b4a1101b99e5f882f)
Target: x86_64-unknown-linux-gnu
Thread model: posix
Build config: +assertions
```

**Comment by steakhal** [comment]
@NagyDonat I think this relates to loops.

**Comment by chrchr-github** [comment]
Goes back to clang 4.0.0: https://godbolt.org/z/YK1Kx931M

**Comment by bozicrHT** [comment]
Here is an even smaller reproducer that I believe is related:

```c
#define NULL ((void *)0)

extern int *g;
void test(int **p)
{
    if (g == NULL || p != &g)
        return;

    clang_analyzer_eval(p == &g); // UNKNOWN
    // Expected: TRUE

    *p = NULL;
    p = NULL;

    clang_analyzer_eval(g == NULL); // FALSE
    // Expected: TRUE

    if (g != NULL)
        *p = NULL; // False positive.
}
```
After taking the false branch of `p != &g`, the analyzer should know that `p == &g`. However, that relation does not appear to be retained in the program state. As a result, the next store through p: `*p = NULL;` is not reflected as an update to g, even though p is known to alias `&g` on this path.

**Comment by NagyDonat** [comment]
This bug is an architectural limitation of the way the analyzer represents symbolic values (and has nothing to do with loops).

Expressions like `head->previous != &head` (or `head->previous == &head`) compare a symbolic pointer (the unknown pointer value that is stored in `head->previous`) and a pointer pointing to a non-symbolic region (`&head`, the address of the memory region declared under the name `head`) -- and the analyzer simply cannot represent the value of such expressions. (See also [the reproducer in godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYMQkgGykHAGTwGTAA5dwAjTGIQAGZfAAdUBUJbBmc3Dy8EpJSBQOCwlkjouItMKxsBIQImYgJ0908fMorU6tqCfNCIqNjfBRq6hszmgY6uwuK%2BgEoLVFdiZHYOAFINAEEVgCYYrBpggGoQuX9/A/ODqAA3VDx0ACppjWm1zY2bu4PkAyNjQzEAJ4ALyixkwVzEECCBBeMQAQq8Pugvj9gH9mLRgaCGO4ACp4NgKABKmCYyAQmHQECRsIRG1eBAB8UpmCoBwGxFc1iOGEwPKwK3hrw5XII/L5KwA7HT1hdxQd7sFVARBTK5SFeQr7vFiODbK4FKrXlKACJG%2BnvW7I76GNH/TEg4jGdDueLGBi8%2BLxCAi7karD3R7m9ZIlG29GAx3O13GaFQwS04U1GzIeX3ClMdDBxFWg6uBi0IIAa3dvIg/r5it5L3p0tecqrWAAtILsDq9fMFAdBSaDh7m62lSqhRbZRc8GyIP3MC2YtghwcwGAe0cTv4a2O5edp7P55hlbv21d9V2VzvW0eT9mNnLz3PL53uzFe8dTsHb7zdwuV6//NnJWao6hrqOoYFyUTUlaG5Smq46ThmyI9j%2Ba7dpKjhSo4BwIV%2B%2B5ikuyGnKh6FoVhpLoIeurHo%2B%2BHPt2WzeAhG5bhcuoEAsDDXpuYa/PaWJOi6LButOXoQNhF6USeiY3hcNo8RifHRoJpZYCJ2wMWRUlcbJdryVGAlunGYn3hJj5IbRamMe%2BMmohGDqgvpsaCKJZEUR2BqLsu5n0ZZI5vFxADuCB0HyzmZh5K7PKhsHMdptkKTiLD4oSJJkhSVKafWzFptOT69ghVnMfmhYMCW05TtWnFZdxOmRqC4KQghuUEeuBWoYBmwARwsy0JwACsvCeBwWikKgnCYQo8yLBKOw8KQBCaF1sxFiAvUaAAdBoACckoABySJKMS9TtXBbFwMQxPonCSLwLASBoGikINw2jRwvAKCAD3zUNXWkHAsAwIgKCoIJwVkBQEBoCD9DRMgsUEJyJV8HQBBRO9EDhAtpDhEEtQApws3Y8wxAAgA8uE2iYNY%2BO8JDbCCCTBZ499pBYOErjAI4Yi0O93C8FgLC2uIzP4Lq1h4FcmA88N%2B6U64KPU%2BQgjlArhbhMQuPOFgmPwwSCsS8Q4RJJgJqYALRjFaA32zFQBjAAoABqeCYP5JPMoNs38IIIhiOwUgyIIigqOozO6Fw%2Bi2iApjGOYqvvZAsyoPElQMDzTZNrxIIHE2AAa2lZxnURNuSlNFlEWe56iXwkJgLOYGzwBrdgqg6gAkinzJiwIvCoPrxB3JL8CzJYlPJ/YDBOC4jR6AEQTdEUvRh4kyTJ8MniLzkycTD00Rh8PncMO0QyT5ku/lCPbSDJ0s%2BTAvFiX6vehjHUW/zzvQ%2BTUsEjdX1A2Yy9ByqB2t4Js3hJDVUuPDfMRZpiXFwIQEgdEYhcGmLwL6WhphLRWg9HqHBrqkFuqtR6f9OBvQ%2BnNBasw/qA3mAQeIctyCUEhvEUGIRWDLEAcA0BBxgCwwOFwbwa1ZqUngX3dAehPbCFEOIP2EjA5qExqHUg/l1bxGpt/Dg/UiHMxeiTOWdCxSoDZBwkBYC84QEgSVGBEBnBQzLtsJBKDyFW1mAhXoEB1F4IIQ9J63cSEWDIWgxapBlqSE2mtTaGhTq9XopILgO1dreElJdDgMRf7aL8YEjBySthpOehkihsx9bJDsJIIAA%3D%3D%3D)).

We have `SymSymExpr` (applying a binary operator to two symbols), `SymIntExpr` (applying a binary operator to a symbol and an integer), `IntSymExpr` (the same with reversed operand order) and `UnarySymExpr` (applying an unary operator to a symbol) – but we don't have something like `SymRegionExpr` so `SimpleSValBuilder::evalBinOpLL` just returns `UnknownVal()` at the very end.

I would caution against introducing `SymRegionExpr` (and its counterpart `RegionSymExpr`) because it would just exacerbate the problem that the `SymExpr` class hierarchy was developed in an ad hoc fashion, without a clear architectural design.

Eventually we should fundamentally redesign the `SymExpr`, `SVal` and `MemRegion` class hierarchies because their ad hoc design and many shortcomings sharply limit the capabilities of the analyzer. Without good foundations, we can only build a heap of hacky workarounds, that are ineffective to develop and maintain.

Based on this I'm leaning towards closing this as wontfix because I don't see a good solution. @steakhal What do you think?

**Comment by steakhal** [comment]
I think your assessment is correct. I'd advocate for NOT redesigning the fundamental class hierarchies.
Continuity is the essence - even if it has problems. Remember for the Chesterton’s fence.
